### PR TITLE
[PT FE] Cannot get length of dynamic dimension

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Fix output tensor creation for bound evaluation when node output has dynamic element type and dynamic shapes.

### Tickets:
 - 102801

### CODE REVIEW

File: src/core/src/bound_evaluate.cpp [hunk 1]
```diff
@@ -80,8 +80,17 @@ ov::Tensor evaluate_bound(const Output<Node>& output, bool is_upper, bool invali
...
+
+                if (out_et.is_dynamic()) {
+                    outputs.emplace_back();
+                } else if (out_shape.is_static()) {
+                    outputs.emplace_back(out_et, out_shape.to_shape());
+                } else if (out_shape.rank().is_static()) {
+                    outputs.emplace_back(out_et, Shape(out_shape.rank().get_length()));
+                } else {
+                    outputs.emplace_back(out_et, Shape{0, std::numeric_limits<size_t>::max()});
+                }
```
Comment (mvafin) [inline src/core/src/bound_evaluate.cpp:93]:
```suggestion
                if (out_shape.is_static()) {
                    outputs.emplace_back(out_et, out_shape.to_shape());
                } else if (out_shape.rank().is_static()) {
                    outputs.emplace_back(out_et, Shape(out_shape.rank().get_length()));
                } else {
                    outputs.emplace_back();
                }
```
This works too and I don't think that you really want to check for dynamic type here, only dynamic rank.

Comment (praasz) [reply src/core/src/bound_evaluate.cpp:93]: Is not possible to create ov::Tensor with dynamic type (default allocator will throw exception), In theory you can have static shape and dynamic type which trigger exception here.

Comment (mvafin) [reply src/core/src/bound_evaluate.cpp:93]: Ah, I see, ok.

File: src/core/src/bound_evaluate.cpp [hunk 2]
```diff
@@ -80,8 +80,17 @@ ov::Tensor evaluate_bound(const Output<Node>& output, bool is_upper, bool invali
             ov::TensorVector outputs;
             for (const auto& out : node->outputs()) {
                 const auto& out_shape = out.get_partial_shape();
-                auto shape = out_shape.is_static() ? out_shape.to_shape() : Shape(out_shape.rank().get_length());
-                outputs.emplace_back(out.get_element_type(), shape);
+                const auto& out_et = out.get_element_type();
+
```
Comment (rkazants) [inline src/core/src/bound_evaluate.cpp:84]: Can we have test for your changes please?

Comment (praasz) [reply src/core/src/bound_evaluate.cpp:84]: Based on reproduction scenario add tests to simulate similar situation when outputs can have dynamic element type or dynamic rank.

**Review by mmikolajcz — APPROVED** [review]

File: src/core/src/bound_evaluate.cpp [hunk 3]
```diff
@@ -80,8 +80,17 @@ ov::Tensor evaluate_bound(const Output<Node>& output, bool is_upper, bool invali
...
+                if (out_et.is_dynamic()) {
+                    outputs.emplace_back();
+                } else if (out_shape.is_static()) {
+                    outputs.emplace_back(out_et, out_shape.to_shape());
+                } else if (out_shape.rank().is_static()) {
+                    outputs.emplace_back(out_et, Shape(out_shape.rank().get_length()));
+                } else {
+                    outputs.emplace_back(out_et, Shape{0, std::numeric_limits<size_t>::max()});
```
Comment (rkazants) [inline src/core/src/bound_evaluate.cpp:92]: not clear why you create Shape like for 1D tensor if the rank is dynamic? Please clarify

File: src/core/tests/bound_evaluate.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        fn_op->set_attrs(attrs);
+        fn_op->set_function(0, body);
+        fn_op->set_function(1, body);
+    }
+
+    std::shared_ptr<op::util::FrameworkNode> fn_op;
+};
+
+// Simulate scenarios in pytorch when eval bounds for node PtFrameworkNode which inherits from FrameworkNode
+TEST_F(EvaluateBoundTest, no_exception_wehen_node_has_output_with_dynamic_rank) {
```
Comment (rkazants) [inline src/core/tests/bound_evaluate.cpp:39]:
```suggestion
TEST_F(EvaluateBoundTest, no_exception_when_node_has_output_with_dynamic_rank) {
```

File: src/core/tests/bound_evaluate.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    fn_op->set_output_type(0, element::i32, PartialShape::dynamic());
+    fn_op->set_output_type(1, element::i32, PartialShape{{1, 4}});
+    fn_op->validate_and_infer_types();
+
+    EXPECT_NO_THROW(evaluate_both_bounds(fn_op));
+}
+
+TEST_F(EvaluateBoundTest, no_exception_wehen_node_has_output_with_dynamic_element_type) {
```
Comment (rkazants) [inline src/core/tests/bound_evaluate.cpp:47]:
```suggestion
TEST_F(EvaluateBoundTest, no_exception_when_node_has_output_with_dynamic_element_type) {
```

File: src/core/src/bound_evaluate.cpp (hunk 3 above)
Comment (rkazants) [reply src/core/src/bound_evaluate.cpp:92]: Or how we (the logic below) will distinguish it from a case of 2d tensor of shape `0, std::numeric_limits<size_t>::max()`

Comment (praasz) [reply src/core/src/bound_evaluate.cpp:92]: Yes true, changed to Shape{0} as is used as special case to convert to dynamic tensor.

Comment (rkazants) [reply src/core/src/bound_evaluate.cpp]: @jane-intel, what do you think how we should handle this case?

Comment (jane-intel) [reply src/core/src/bound_evaluate.cpp]: I think that if output shape of the operation is yet to be known -- we may let it be scalar and perform proper memory allocation later -- during evaluate method we would collect exact number of elements in the output and will set output shape. So, I agree with the solution.

**Review by jane-intel — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
is it valid to port this change to 2022.3?

**Pull request merged**